vector/src/main/java/im/vector/app/features/onboarding/ftueauth/FtueAuthCombinedServerSelectionFragment.kt

```diff
@@ -36,67 +40,113 @@ import im.vector.app.features.onboarding.OnboardingAction
 import im.vector.app.features.onboarding.OnboardingFlow
 import im.vector.app.features.onboarding.OnboardingViewEvents
 import im.vector.app.features.onboarding.OnboardingViewState
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
 import org.matrix.android.sdk.api.failure.isHomeserverUnavailable
+import reactivecircus.flowbinding.android.view.clicks
 import javax.inject.Inject
 
-class FtueAuthCombinedServerSelectionFragment @Inject constructor() : AbstractFtueAuthFragment<FragmentFtueServerSelectionCombinedBinding>() {
+class FtueAuthCombinedServerSelectionFragment @Inject constructor(
+        private val stringProvider: StringProvider,
+) : AbstractFtueAuthFragment<FragmentFtueServerSelectionCombinedBinding>() {
+
+    private lateinit var controller: Controller
 
     override fun getBinding(inflater: LayoutInflater, container: ViewGroup?): FragmentFtueServerSelectionCombinedBinding {
-        return FragmentFtueServerSelectionCombinedBinding.inflate(inflater, container, false)
+        return FragmentFtueServerSelectionCombinedBinding.inflate(inflater, container, false).also {
+            controller = Controller(viewLifecycleOwner, it, stringProvider, errorFormatter)
+        }
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        setupViews()
-    }
-
-    private fun setupViews() {
-        views.chooseServerRoot.realignPercentagesToParent()
-        views.chooseServerToolbar.setNavigationOnClickListener {
-            viewModel.handle(OnboardingAction.PostViewEvent(OnboardingViewEvents.OnBack))
-        }
-        views.chooseServerInput.associateContentStateWith(button = views.chooseServerSubmit, enabledPredicate = { canSubmit(it) })
-        views.chooseServerInput.setOnImeDoneListener {
-            if (canSubmit(views.chooseServerInput.content())) {
-                updateServerUrl()
+        controller.listener = object : Controller.Listener {
```

Comment:
I would prefer the Fragment to implement `Controller.Listener`. Less indentation, and also less change on this PR.